src/Sprites/ImageHeader.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include "BmpCompression.h"
+#include <cstdint>
+#include <array>
+
+#pragma pack(push, 1) // Make sure structures are byte aligned
+
+struct ImageHeader
+{
+	static const std::array<uint16_t, 6> validBitCounts;
+	static const std::array<uint16_t, 3> indexedBitCounts;
+
+	static bool IsValidBitCount(uint16_t bitCount);
+	static bool IsIndexedBitCount(uint16_t bitCount);
+
+	static void CheckValidBitCount(uint16_t bitCount);
+	static void CheckIndexedBitCount(uint16_t bitCount);
+
+	static ImageHeader Create(int32_t width, int32_t height, uint16_t bitCount)
+	{
+		CheckValidBitCount(bitCount);
+
+		return ImageHeader{
+			40,
+			width,
+			height,
+			1,
+			bitCount,
+			BmpCompression::Uncompressed,
+			0,
+			0,
+			0,
+			0,
+			0 };
```

Comment:
Agree on fixing the hardcoded 40. 

After thinking about it, I think we could just set static const variables as part of the ImageHeader. It would serve for self-documentation of what is default and if someone wants to create an ImageHeader not using our static Create function, they could still pull some of the default values.

``` C++
static const uint16_t defaultPlanes = 1;
static const BmpCompression defaultCompression = CmpCompression::Uncompressed;
static const uint32_t  defaultXResolution = 0;
static const uint32_t defaultYResolution = 0;
static const uint32_t defaultUsedColorMapEntries = 0;
static const uint32_t defaultImportantColorCount = 0;
```

Reply:
Ahh, that's not so bad. For `BmpCompression::Uncompressed` I think we can use it as is. The name is self documenting.

For `defaultUsedColorMapEntries`, I believe the default would be `255` for 8-bit palettized bitmaps. I'm not sure what happens if you use `0` in that context, though it might also be fine. I would expect to see `0` for true color bitmaps, `2` for 1-bit bitmaps, and `16` for 4-bit bitmaps. I might also expect `defaultImportantColorCount` to match `defaultUsedColorMapEntries`. Again though, using `0` might be valid.

Reply:
My memory from reading the specs is that if you use a full palette, you can use 0 for UsedColorMapEntries (which is what MS Paint does).

> In practice it is very uncommon to run across images that have anything other than zero for this parameter. The reason is that it is a simple matter to go ahead and set all of the unused colors to all zero (black) -- or some other color -- and store all 2N of them in the Color Table. The space penalty is negligible since the file size is dominated by the pixel data for all but the smallest images.

ImportantColorCount can be 0 if you are not defining any colors as more important than others. While I understand that you can designate important vs non-important colors, it seems silly to implement as the application would have to chose how to display the non-important colors on its own.

> If this parameter is equal to zero, then all colors in the Color Table are to be considered important. 

-Brett

Reply:
Ok, sounds good then. Let's just set the values to 0 and not worry about it.

Now that I think about it, `defaultImportantColorCount` was really only relevant back in the day when people used 256 color desktop settings. In that case, if you were using multiple programs at the same time, you might need to make do with a more limited set of colors. Hence you could map the less important colors to different available system colors, or use dithering to approximate the intended color. It never struck me as a very workable system. How do you even determine how many colors are actually important? At any rate, it's a long obsolete consideration now. Static default values will be good.